mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -0,0 +1,334 @@
+    modelSubsetWithDistCol.sort(distCol).limit(numNearestNeighbors)
+  }
+
+  /**
+   * Overloaded method for approxNearestNeighbors. Use Single Probing as default way to search
+   * nearest neighbors and "distCol" as default distCol.
+   */
+  @Since("2.1.0")
+  def approxNearestNeighbors(
+      @Since("2.1.0") dataset: Dataset[_],
+      @Since("2.1.0") key: Vector,
+      @Since("2.1.0") numNearestNeighbors: Int): Dataset[_] = {
+    approxNearestNeighbors(dataset, key, numNearestNeighbors, true, "distCol")
+  }
+
+  /**
+   * Preprocess step for approximate similarity join. Transform and explode the outputCol to
+   * explodeCols.
+   * @param dataset The dataset to transform and explode.
+   * @param explodeCols The alias for the exploded columns, must be a seq of two strings.
+   * @return A dataset containing idCol, inputCol and explodeCols
+   */
+  @Since("2.1.0")
+  private[this] def processDataset(dataset: Dataset[_], explodeCols: Seq[String]): Dataset[_] = {
+    require(explodeCols.size == 2, "explodeCols must be two strings.")
+    val vectorToMap: UserDefinedFunction = udf((x: Vector) => x.asBreeze.iterator.toMap,
+      MapType(DataTypes.IntegerType, DataTypes.DoubleType))
+    val modelDataset: DataFrame = if (!dataset.columns.contains($(outputCol))) {
+      transform(dataset)
+    } else {
+      dataset.toDF()
+    }
+    modelDataset.select(col("*"), explode(vectorToMap(col($(outputCol)))).as(explodeCols))
+  }
+
+  /**
+   * Recreate a column using the same column name but different attribute id. Used in approximate
+   * similarity join.
+   * @param dataset The dataset where a column need to recreate
+   * @param colName The name of the column to recreate
+   * @param tmpColName A temporary column name which does not conflict with existing columns
+   * @return
+   */
+  @Since("2.1.0")
+  private[this] def recreateCol(
+      @Since("2.1.0") dataset: Dataset[_],
+      @Since("2.1.0") colName: String,
+      @Since("2.1.0") tmpColName: String): Dataset[_] = {
+    dataset
+      .withColumnRenamed(colName, tmpColName)
+      .withColumn(colName, col(tmpColName))
+      .drop(tmpColName)
+  }
+
+  /**
+   * Join two dataset to approximately find all pairs of records whose distance are smaller
+   * than the threshold.
+   * @param datasetA One of the datasets to join
+   * @param datasetB Another dataset to join
+   * @param threshold The threshold for the distance of record pairs
+   * @param distCol Output column for storing the distance between each result record and the key
+   * @return A joined dataset containing pairs of records. A distCol is added to show the distance
+   *         between each pair of records.
+   */
+  @Since("2.1.0")
+  def approxSimilarityJoin(
+      @Since("2.1.0") datasetA: Dataset[_],
+      @Since("2.1.0") datasetB: Dataset[_],
+      @Since("2.1.0") threshold: Double,
+      @Since("2.1.0") distCol: String): Dataset[_] = {
+
+    val explodeCols = Seq("lsh#entry", "lsh#hashValue")
```

Comment:
Yes, I used this kind of code in early implementations:
`df1.join(df2, shareHashBucket(df1($(outputCol)), df2($(outputCol))))`
However, it seems this is broadcast join + filter. I am wondering if explode is more feasible since it's doing hash join?